OpenRA.Mods.Common/Traits/Buildings/Gate.cs

```diff
@@ -0,0 +1,119 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System.Linq;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Will open and be passable for actors that appear friendly when there are no enemies in range.")]
+	public class GateInfo : BuildingInfo, ITraitInfo
+	{
+		public readonly string OpeningSound = null;
+		public readonly string ClosingSound = null;
+
+		[Desc("Ticks until the gate closes.")]
+		public readonly int TicksToStayOpen = 50;
+
+		[Desc("Ticks until the gate is considered open.")]
+		public readonly int TicksUntilOpen = 10;
+
+		public override object Create(ActorInitializer init) { return new Gate(init, this); }
+	}
+
+	public class Gate : Building, ITick, ITemporaryBlocker, INotifyBlockingMove, ISync
+	{
+		readonly GateInfo info;
+		readonly Actor self;
+
+		public readonly int OpenPosition;
+		[Sync] public int Position { get; private set; }
+		int desiredPosition;
+		int remainingOpenTime;
+
+		public Gate(ActorInitializer init, GateInfo info)
+			: base(init, info)
+		{
+			this.info = info;
+			this.self = init.Self;
+			OpenPosition = info.TicksUntilOpen;
+		}
+
+		void ITick.Tick(Actor self)
+		{
+			if (self.IsDisabled() || !BuildComplete)
+				return;
+
+			if (desiredPosition < Position)
+			{
+				// Gate was fully open
+				if (Position == OpenPosition)
+				{
+					Game.Sound.Play(info.ClosingSound, self.CenterPosition);
+					self.World.ActorMap.AddInfluence(self, this);
+				}
+
+				Position--;
+			}
+			else if (desiredPosition > Position)
+			{
+				// Gate was fully closed
+				if (Position == 0)
+					Game.Sound.Play(info.OpeningSound, self.CenterPosition);
+
+				Position++;
+
+				// Gate is now fully open
+				if (Position == OpenPosition)
+				{
+					self.World.ActorMap.RemoveInfluence(self, this);
+					remainingOpenTime = info.TicksToStayOpen;
+				}
+			}
+
+			if (Position == OpenPosition)
+			{
+				if (IsBlocked())
+					remainingOpenTime = info.TicksToStayOpen;
+				else if (--remainingOpenTime <= 0)
+					desiredPosition = 0;
+			}
+		}
+
+		bool ITemporaryBlocker.IsBlocking(Actor self, CPos cell)
+		{
+			return Position != OpenPosition && FootprintUtils.Tiles(self).Contains(cell);
```

Comment:
It would be worth implementing `INotifyAddedToWorld` and then caching `FootprintUtils.Tiles(self)`, because that does a bunch of work and this gets called a lot.